src/Fixer/Phpdoc/PhpdocParamOrderFixer.php

```diff
@@ -0,0 +1,208 @@
+
+        $params = [];
+        for (
+            $i = $tokens->getNextTokenOfKind($paramBlockStart, [[T_VARIABLE]]);
+            null !== $i && $i < $paramBlockEnd;
+            $i = $tokens->getNextTokenOfKind($i, [[T_VARIABLE]])
+        ) {
+            $params[] = $tokens[$i];
+        }
+
+        return $params;
+    }
+
+    /**
+     * Sorts the param annotations according to the function parameters.
+     *
+     * @param Token   $docBlockToken
+     * @param Token[] $funcParamTokens
+     *
+     * @return DocBlock
+     */
+    private function sortDocBlockParamAnnotations(Token $docBlockToken, array $funcParamTokens)
+    {
+        $doc = new DocBlock($docBlockToken->getContent());
+        $paramAnotations = $doc->getAnnotationsOfType('param');
+
+        $validParams = [];
+        foreach ($funcParamTokens as $funcParam) {
+            $identifier = $funcParam->getContent();
+            $indices = $this->findParamAnnotationByIdentifier($paramAnotations, $identifier);
+
+            // Found an excactly matching @param annotation
+            if (is_array($indices)) {
+                foreach ($indices as $index) {
+                    $validParams[$index] = $paramAnotations[$index]->getContent();
+                }
+            }
+        }
+
+        // Detect superfluous annotations
+        /** @var Annotation[] $invalidParams */
+        $invalidParams = array_diff_key($paramAnotations, $validParams);
+        $invalidParams = array_values($invalidParams);
+
+        // Append invalid parameters to the (ordered) valid ones
+        $orderedAnnotations = array_values($validParams);
+        foreach ($invalidParams as $i => $params) {
+            $orderedAnnotations[$i + count($validParams)] = $params->getContent();
+        }
+
+        // Rewrite the param annotations in order
+        foreach ($paramAnotations as $i => $docAnnotation) {
+            $docAnnotation->remove();
+            $doc
+                ->getLine($docAnnotation->getStart())
+                ->setContent($orderedAnnotations[$i]);
+        }
+
+        return $doc;
+    }
+
+    /**
+     * Returns the indices of the lines of a specific parameter annotation.
+     *
+     * @param Annotation[] $docParams
+     * @param string       $identifier
+     *
+     * @return null|array
+     */
+    private function findParamAnnotationByIdentifier(array $docParams, $identifier)
+    {
+        $inBlock = false;
+        $blockMatch = false;
+        $blockIndices = [];
+
+        $paramRegex = sprintf(
+            '/\*\s*@param\s*[\w%s]*\s*\$\b%s\b/',
+            preg_quote('[]|\?'),
+            substr($identifier, 1) // remove starting `$`
+        );
+
+        foreach ($docParams as $i => $param) {
+            $blockStart = Preg::match('/\s*{\s*$/', $param->getContent());
+            $blockEnd = Preg::match('/\s*}\s*$/', $param->getContent());
+
+            if ($blockStart) {
+                $inBlock = true;
```

Comment:
The fact that you only do true/false for `$inBlock` suggests that this still breaks for a double-nested (or more) doc block. Nesting could happen infinitely deep, and params in a nested block could have the same name as one block higher in the chain. I don't know if this is a case we want to support, but the spec allows it.

Reply:
You are right. I've adapted the code to support deeply nested Docblocks and added tests.